### ISSUE TITLE
quality-control metrics in R: extremely low counts of genes per cell

### ISSUE THREAD

**Issue by Aintzane99** [COMBINE-lab/alevin-fry#57]
Hello, 

I am testing alevin-fry on some public sc-RNAseq data (GEO accession: GSE81076). The issue that I have is that when I process the gene count x cell matrix in R and perform per-cell metrics, I get extremely low counts of genes per cell. I repeated the analysis several times changing some parameters but I don't seem to grasp the error. 

The data has 18 samples, I am now processing only one of them to try to understand where the problem lies. Here is a detailed explanation of what I am doing:

- Characteristics of the data that I'm using:  paired-end, forward stranded, read 1 is composed of 8 first bases corresponding to the cell-barcode followed by 4 bases of the UMI sequence. Read 2 is just the transcript sequence. 

- First, I built the salmon/alevin splici index following the tutorial "Resolving the splicing origins of UMIs to improve the specificity of single-cell RNA-seq transcriptome mapping with alevin-fry"

- Mapping the reads:
salmon alevin -i ../alevin_index/grch38_splici_idx -p 10 -l ISF --bc-geometry 1[1-8] --umi-geometry 1[9-12] --read-geometry 2[1-end] --rad -1 73_1.fastq.gz -2 73_2.fastq.gz -o 73_SAM_pbmc_10k_v3_map --tgMap t2g.tsv --forceCells 96

- Permit-list:
alevin-fry generate-permit-list -d fw --force-cells 96 -i 73_SAM_pbmc_10k_v3_map -o 73_SAM_pbmc_10k_v3_quant

- Collating the file:
alevin-fry collate -t 10 -i 73_SAM_pbmc_10k_v3_quant -r 73_SAM_pbmc_10k_v3_map

- UMI quantification:
alevin-fry quant -t 10 -i 73_SAM_pbmc_10k_v3_quant -o 73_SAM_pbmc_10k_v3_quant_res --tg-map ../tg2_3col.tsv --resolution cr-like-em --use-mtx 

- Getting the SA gene-by-cell matrix:
I used the RScript provided in the tutorial to generate the SingleCellExperiment object. 

Then, I extract the counts assay from the generated SCE object and take a look at it by converting it into a matrix. This is what is looks like:
![image](https://user-images.githubusercontent.com/92913155/169829573-d0d206e5-f7fc-4cb1-98a9-c8b7b1b79e00.png)

As you can see, the gene counts are not whole numbers. After, I perform some quality-control metrics also in R, obtaining the following results for the gene counts per cell:

summary(per.cell$detected)
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#    3.00   31.00   61.00   63.12   92.00  154.00

I tried with different samples but all of them yield very low count results. The percented of mapped reads is 42%, not too high, but I also obtained a similar percentage with other mapping tools, so I suspect that the error could be in the downstream steps. Please, let me know if there is something wrong with my code.

Thanks in advance,

Aintzane.

**Comment by rob-p** [comment]
Hi @Aintzane99,

  We'd be happy to look into this further.  What technology is this — the data seem quite old (the UMIs are really short!).  One thing to try is, in your permit-list step, try:

```
alevin-fry generate-permit-list -d both --force-cells 96 -i 73_SAM_pbmc_10k_v3_map -o 73_SAM_pbmc_10k_v3_quant
```
 
or 

```
alevin-fry generate-permit-list -d rc --force-cells 96 -i 73_SAM_pbmc_10k_v3_map -o 73_SAM_pbmc_10k_v3_quant
```

As that orientation would be consistent with the library type you passed, which expects the first read (i.e. the technical read) to come from the forward strand, implying the second read comes from the rc strand.

Best,
Rob

**Comment by rob-p** [comment]
Hi @Aintzane99,

Indeed, this is almost certainly the issue.  Consider doing the following:

```
cut -f ../tg2_3col.tsv > ../t2g_2col.tsv
```

And then just changing your last step for quantification to be : 

```
alevin-fry quant -t 10 -i 73_SAM_pbmc_10k_v3_quant -o 73_SAM_pbmc_10k_v3_quant_res --tg-map ../tg2_2col.tsv --resolution parsimony --use-mtx
```

Now you will get (as expected) *much* higher counts for genes per cell.  Of course, since the unspliced and spliced annotations have been combined here, you won't get a "USA" mode output and your total gene expression signal will be the combination of both spliced and unspliced reads.

The reason this happens is that the `cr-like` and `cr-like-em` methods, following their namesake (CellRanger), make the assumption that a `(umi, barcode)` pair should be unique.  This assumption can be rather reasonable when UMIs are long (e.g. as in the case of many modern technologies), but doesn't make sense when the UMIs are of length 4.  Support for USA mode for `parsimony` resolution is planned, but the other think that should be considered in data like this is whether or not one should consider edit distance 1 UMI pairs to potentially come from the same molecule.  Again, in `parsimony` resolution mode, for longer UMIs, it is assumed that it is possible for two UMIs that differ by only a single edit to have arisen from the same molecule (with the edit being induced by PCR or sequencing error).  However, it's not clear if this should be considered the case with such short UMIs.  What are your thoughts on this?  On one hand, we think there is value in having the `parsimony` resolution support USA mode, and even in adding other resolution modes.  On the other hand, we likely want to limit the effort we expend on supporting protocols with very short UMIs that are unlikely to be used much anymore.  I'm looping in @DongzeHE on this as well so he's aware of the conversation.

Best,
Rob

**Comment by rob-p** [comment]
Hi @Aintzane99,

  Thanks for the follow up.  Indeed, these warnings are precisely due to the very short UMI length being used in these protocols.  Specifically, the probability of a pair of UMIs colliding (mapping to an overlapping set of transcripts) scales inversely and geometrically with the UMI length.  The collision probability is exacerbated by the fact that 1-edit neighbors are considered to be resolvable to the same UMI if the map to the same gene and have a sufficiently different count.  

However, this is likely to be the case with almost any UMI resolution method I am aware of, at least with UMIs this short and with the coverage of data you are seeing in these samples — as there are only 256 distinct values a UMI can take on.

That being said, we've released `alevin-fry` 0.6.0 and added some more features that may be helpful in this case.  In particular, the `--parsimony` and `--parsimony-em` methods now work with USA mode (so you don't need to do that hack with the t2g file).  Also, we added a (hidden) option `--umi-edit-dist` that you can set to 0 when using the `--parsimony` and `--parsimony-em` modes. This requires UMIs to be *identical* for merging, rather than just being within a Hamming distance of 1.  There are some other additions too that you can see in the [release notes](https://github.com/COMBINE-lab/alevin-fry/releases/tag/v0.6.0).

Overall, I'd say these are your best bet in trying to analyze this data with `alevin-fry` or with any other modern pipelines for single-cell preprocessing.  However, I'll certainly reiterate the caveat that these UMIs are *very* short, and therefore you should maintain a healthy level of skepticism of the results you see given the substantial probability that UMI over-collapsing will occur, at least for highly-expressed genes across cells.

I hope this helps.  Please feel free to re-open or open another issue if you have follow up questions.

Best,
Rob
